File: entities/ranger.py
# entities/ranger.py
import math
import pygame
import random
from .base_entity import Entity
from settings import *


class Ranger(Entity):
    def __init__(self, game, x, y):
        super().__init__(game, x, y, (0, 255, 255), "diamond")
        self.speed = 2.0
        self.state = "positioning"
        self.charge_timer = 0
        self.ideal_range = 300

        self.is_firing = False
        self.fire_frame = 0

    def update(self):
        super().update()

        if self.is_firing:
            self.fire_frame -= 1
            if self.fire_frame <= 0:
                self.is_firing = False
                self.state = "positioning"
            return

        px, py = self.game.player.pos
        dist_p = self.get_distance_to(px, py)

        if self.state == "positioning":
            self.color = (0, 255, 255)

            if dist_p < self.ideal_range - 50:
                angle_away = math.atan2(self.y - py, self.x - px)
                self.angle = angle_away
                self.move(math.cos(self.angle) * self.speed, math.sin(self.angle) * self.speed)

            elif dist_p > self.ideal_range + 50:
                angle_to = math.atan2(py - self.y, px - self.x)
                self.angle = angle_to
                self.move(math.cos(self.angle) * self.speed, math.sin(self.angle) * self.speed)

            else:
                self.state = "aiming"
                self.charge_timer = 60

        elif self.state == "aiming":
            self.color = (255, 255, 255)
            self.charge_timer -= 1

            self.angle = math.atan2(py - self.y, px - self.x)

            if self.charge_timer <= 0:
                self.state = "firing"
                self.fire_shot()

    def fire_shot(self):
        self.is_firing = True
        self.fire_frame = 5
        self.color = (255, 0, 0)
        # Ici on pourrait réduire la vie du joueur
        print("Ranger fired!")

    def draw_2d(self, surface, offset_x, offset_y, zoom):
        super().draw_2d(surface, offset_x, offset_y, zoom)

        if self.is_firing:
            sx = self.x * zoom + offset_x
            sy = self.y * zoom + offset_y

            range_len = 600 * zoom
            ex = sx + math.cos(self.angle) * range_len
            ey = sy + math.sin(self.angle) * range_len

            pygame.draw.line(surface, (255, 255, 0), (sx, sy), (ex, ey), 2)

File: src/entities/ranger.rs
// src/entities/ranger.rs
use crate::map::MapHandler;
use super::entity_base::EntityBase;

#[derive(Clone, Copy, PartialEq)]
pub enum RangerState { Positioning, Aiming, Firing }

pub fn update_ranger(
    base: &mut EntityBase,
    state: &mut RangerState,
    charge_timer: &mut i32,
    ideal_range: &f32,
    is_firing: &mut bool,
    fire_frame: &mut i32,
    player_pos: (f32, f32),
    map: &mut MapHandler,
    dt: f32,
) {
    let speed = base.speed * dt * 60.0;
    if *is_firing {
        *fire_frame -= 1;
        if *fire_frame <= 0 {
            *is_firing = false;
            *state = RangerState::Positioning;
        }
        return;
    }

    let (px, py) = player_pos;
    let dist_p = base.dist_to(px, py);

    match state {
        RangerState::Positioning => {
            base.color = [0.0, 200.0/255.0, 200.0/255.0];
            if dist_p < *ideal_range - 50.0 {
                let away = (base.y - py).atan2(base.x - px);
                base.angle = away;
                base.move_pos(away.cos() * speed, away.sin() * speed, map);
            } else if dist_p > *ideal_range + 50.0 {
                let to = (py - base.y).atan2(px - base.x);
                base.angle = to;
                base.move_pos(to.cos() * speed, to.sin() * speed, map);
            } else {
                *state = RangerState::Aiming;
                *charge_timer = 60;
            }
        }
        RangerState::Aiming => {
            *charge_timer -= 1;
            base.angle = (py - base.y).atan2(px - base.x);
            if *charge_timer % 10 < 5 {
                base.color = [1.0, 1.0, 1.0];
            } else {
                base.color = [0.0, 200.0/255.0, 200.0/255.0];
            }
            if *charge_timer <= 0 {
                *state = RangerState::Firing;
                *is_firing = true;
                *fire_frame = 5;
                base.color = [1.0, 50.0/255.0, 50.0/255.0];
            }
        }
        RangerState::Firing => {}
    }
}
